### DISCUSSION TITLE
A conditional step with a previous output and failure won't run

### DISCUSSION THREAD

**Discussion by utgwkk** [community/community#24941] [Actions]
<p>I wrote the following workflow file to invoke upload-artifact if do_test fails and hello.txt is outputted due to do_test. But contrary to what I want, the workflow always skips the last step, shown <a href="https://github.com/utgwkk/sandbox-actions-if/runs/678040046" rel="nofollow noopener">here</a>. Is there any solution?</p>
<pre><code>name: test
on:
  push:

jobs:
  test:
    name: annotate
    runs-on: ubuntu-latest
    steps:
# Instead of actual tests, write a file and return non-zero status to simulate the situation.
      - id: do_test
        run: |
          echo hello &gt; hello.txt
          false

      - id: foo
        run: |
          if [-e 'hello.txt']; then
            echo "::set-output name=exists::true"
          fi
        if: failure()

      - name: check steps.foo.outputs.exists
        run: echo ${{ steps.foo.outputs.exists }}
        if: always()

      - uses: actions/upload-artifact@v2
        with:
          name: hello
          path: hello.txt
        if: steps.foo.outputs.exists
</code></pre>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/utgwkk">@utgwkk</a> ,</p>
<p>The upload-artifact step is not executed due to previous step ‘do_test’ failed. You need to add failure() expression check as well. Code sample as below:</p>
<pre><code class="lang-auto">    steps:
          - id: do_test
            run: |
              echo hello &gt; hello.txt
              false
          - id: foo
            run: |
              if [-e 'hello.txt']; then
                echo "::set-output name=exists::true"
              fi
            if: failure()
    
          - name: check steps.foo.outputs.exists
            run: echo ${{ steps.foo.outputs.exists }}
            if: always()
    
          - uses: actions/upload-artifact@v2
            with:
              name: hello
              path: hello.txt
            if: failure() &amp;&amp; (steps.foo.outputs.exists == 'true')
</code></pre>
<p>Or you can use ‘continue-on-error’ to make step1 always succeeds. Code as below:</p>
<pre><code class="lang-auto">    steps:
          - id: do_test
            continue-on-error: true
            run: |
              echo hello &gt; hello.txt
              false
          - id: foo
            run: |
              if [-e 'hello.txt']; then
                echo "::set-output name=exists::true"
              fi
          - name: check steps.foo.outputs.exists
            run: echo ${{ steps.foo.outputs.exists }}
    
          - uses: actions/upload-artifact@v2
            with:
              name: hello
              path: hello.txt
            if: steps.foo.outputs.exists == 'true'
</code></pre>
<p>Thanks.</p>
